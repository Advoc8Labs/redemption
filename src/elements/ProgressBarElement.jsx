import React from 'react';
import ProgressBar from 'react-bootstrap/ProgressBar';

function ProgressBarElement({ element, curDepth, maxDepth }) {
    const styles = element.applyStyles;

    const percent =
        element.progress || Math.round((100 * curDepth) / (maxDepth + 1));
    const progressBarElements = [
        <ProgressBar
            key='progress'
            style={{
                height: '0.4rem',
                width: '100%',
                borderRadius: 0,
                display: 'flex',
                backgroundColor: '#e9ecef'
            }}
            css={{
                '.progress-bar': {
                    margin: '0 0 0 0 !important',
                    transition: 'width 0.6s ease',
                    ...styles.getTarget('bar')
                }
            }}
            now={percent}
        />
    ];
    const completionPercentage = (
        <div
            key='completionPercentage'
            style={{ width: '100%', textAlign: 'center' }}
        >
            {`${percent}% completed`}
        </div>
    );
    if (element.styles.percent_text_layout === 'top') {
        progressBarElements.splice(0, 0, completionPercentage);
    } else if (element.styles.percent_text_layout === 'bottom') {
        progressBarElements.splice(1, 0, completionPercentage);
    }

    return (
        <div
            key='progress-bar'
            css={{
                display: 'flex',
                flexDirection: 'column',
                ...styles.getLayout(),
                ...styles.getTarget('container')
            }}
        >
            <div
                css={{
                    display: 'flex',
                    flexDirection: 'column',
                    ...styles.getTarget('barContainer')
                }}
            >
                {progressBarElements}
            </div>
        </div>
    );
}

export default ProgressBarElement;
