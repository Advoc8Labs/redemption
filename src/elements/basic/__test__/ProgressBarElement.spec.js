import React from 'react';
import { create, act } from 'react-test-renderer';
import ProgressBarElement from '../ProgressBarElement';
import ApplyStyles from '../../styles';

describe('ProgressBarElement', () => {
  it('renders an empty progress bar element', async () => {
    // Arrange
    const props = {
      element: {
        column_index: 0,
        row_index: 0,
        column_index_end: 0,
        row_index_end: 0,
        styles: {
          border_top_color: '2954af',
          border_right_color: '2954af',
          border_bottom_color: '2954af',
          border_left_color: '2954af',
          background_color: '2954af',
          layout: 'center'
        },
        mobile_styles: {}
      }
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let progressBar;
    act(() => {
      progressBar = create(<ProgressBarElement {...props} />);
    });

    // Assert
    expect(progressBar).toMatchSnapshot();
  });

  it('renders a non-empty progress bar element', async () => {
    // Arrange
    const props = {
      element: {
        column_index: 0,
        row_index: 0,
        column_index_end: 0,
        row_index_end: 0,
        styles: {
          border_top_color: '2954af',
          border_right_color: '2954af',
          border_bottom_color: '2954af',
          border_left_color: '2954af',
          background_color: '2954af',
          layout: 'center'
        },
        mobile_styles: {}
      },
      curDepth: 20,
      maxDepth: 100
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let progressBar;
    act(() => {
      progressBar = create(<ProgressBarElement {...props} />);
    });

    // Assert
    expect(progressBar).toMatchSnapshot();
  });

  it('renders a progress bar with top positioned percent text', async () => {
    // Arrange
    const props = {
      element: {
        column_index: 0,
        row_index: 0,
        column_index_end: 0,
        row_index_end: 0,
        styles: {
          layout: 'center',
          percent_text_layout: 'top'
        },
        mobile_styles: {}
      },
      curDepth: 20,
      maxDepth: 100
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let progressBar;
    act(() => {
      progressBar = create(<ProgressBarElement {...props} />);
    });

    // Assert
    expect(progressBar).toMatchSnapshot();
  });

  it('renders a progress bar with bottom positioned percent text', async () => {
    // Arrange
    const props = {
      element: {
        column_index: 0,
        row_index: 0,
        column_index_end: 0,
        row_index_end: 0,
        styles: {
          layout: 'center',
          percent_text_layout: 'bottom'
        },
        mobile_styles: {}
      },
      curDepth: 20,
      maxDepth: 100
    };
    props.applyStyles = new ApplyStyles(props.element, []);

    // Act
    let progressBar;
    act(() => {
      progressBar = create(<ProgressBarElement {...props} />);
    });

    // Assert
    expect(progressBar).toMatchSnapshot();
  });
});
