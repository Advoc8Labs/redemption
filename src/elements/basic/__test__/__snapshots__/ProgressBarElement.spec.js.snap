// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBarElement renders a non-empty progress bar element 1`] = `
<div
  className="css-taptzq-ProgressBarElement"
>
  <div
    className="css-v1tuzq-progressBarElements progress"
    style={
      Object {
        "backgroundColor": "#e9ecef",
        "borderRadius": 0,
        "display": "flex",
        "height": "0.4rem",
        "width": "100%",
      }
    }
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;

exports[`ProgressBarElement renders a progress bar with bottom positioned percent text 1`] = `
<div
  className="css-taptzq-ProgressBarElement"
>
  <div
    className="css-v1tuzq-progressBarElements progress"
    style={
      Object {
        "backgroundColor": "#e9ecef",
        "borderRadius": 0,
        "display": "flex",
        "height": "0.4rem",
        "width": "100%",
      }
    }
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "20%",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    20% completed
  </div>
</div>
`;

exports[`ProgressBarElement renders a progress bar with top positioned percent text 1`] = `
<div
  className="css-taptzq-ProgressBarElement"
>
  <div
    style={
      Object {
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    20% completed
  </div>
  <div
    className="css-v1tuzq-progressBarElements progress"
    style={
      Object {
        "backgroundColor": "#e9ecef",
        "borderRadius": 0,
        "display": "flex",
        "height": "0.4rem",
        "width": "100%",
      }
    }
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;

exports[`ProgressBarElement renders an empty progress bar element 1`] = `
<div
  className="css-taptzq-ProgressBarElement"
>
  <div
    className="css-v1tuzq-progressBarElements progress"
    style={
      Object {
        "backgroundColor": "#e9ecef",
        "borderRadius": 0,
        "display": "flex",
        "height": "0.4rem",
        "width": "100%",
      }
    }
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={50}
      className="progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "50%",
        }
      }
    />
  </div>
</div>
`;
